src/mlpack/methods/ann/loss_functions/poisson_nll_loss.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file poisson_nll_loss.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Definition of the Poisson Negative Log Likelihood class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_HPP
+#define MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Implementation of the poisson negative log likelihood layer. The poisson
+ * negative log likelihood layer expects input for each class.
+ * The layer also expects a class index, in the range between 1 and the
+ * number of classes, as target when calling the Forward function.
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+  typename InputDataType = arma::mat,
+  typename OutputDataType = arma::mat
+>
+class PoissonNLLLoss
+{
+ public:
+  /**
+   * Create the PoissonNLLLoss object.
+   *
+   * @param logInput If true the loss is computed as exp(input) - target*input,
+   *        if false then the loss is input - target * log(input + eps)
+   * @param full Boolean value that tell if the full loss has to be calculated
+   *        i.e to add Stirling approximation term
+   * @param eps A small value to prevent log(0) when logInput = False
+   * @param reduce Specifies which reduction to be applied to the output
+   *        i.e sum'|'mean'. The corresponding boolean values are 0 and 1.
+   */
+  PoissonNLLLoss(
+    const bool logInput = true,
+    const bool full = false,
+    const double eps = 1e-08,
+    const bool reduce = 1);
```

Comment:
Looking at the parameter description, should we call this `reduction` instead of `reduce`?

Reply:
ok.